### PR TITLE
PRESIDECMS-2030 flag record feature in data manager

### PR DESCRIPTION
**Pull request by pixl8-brayden**

### CODE REVIEW

File: system/services/presideObjects/PresideObjectService.cfc [hunk 1]
```diff
@@ -1634,6 +1634,41 @@ component displayName="Preside Object Service" {
...
+	public string function getFlagField( required string objectName ) {
+		var flagEnabled = getObjectAttribute( arguments.objectName, "flagEnabled", "" );
+
+		if ( IsBoolean( flagEnabled ) && flagEnabled ) {
+			return getObjectAttribute( arguments.objectName, "flagField", "recordFlagged" );
+		}
+
+		return "";
+	}
+
```
Comment (DominicWatson) [inline system/services/presideObjects/PresideObjectService.cfc:1652]: I suggest adding a helper here: `isFlaggingEnabled( objectName )` - this should be used over `Len( getFlagField() )` as a more explicit and readable check in handlers, etc.

File: system/handlers/admin/DataManager.cfc [hunk 2]
```diff
@@ -489,6 +489,40 @@ component extends="preside.system.base.AdminHandler" {
 		}
 	}
 
+	public void function flagRecordAction( event, rc, prc, args={} ) {
+		var objectName = rc.object ?: "";
+		var recordId   = rc.id     ?: "";
+		var resultView = rc.result ?: "";
+		var successUrl = event.buildAdminLink( objectName=objectName );
+		var updated    = false;
+
```
Comment (DominicWatson) [inline system/handlers/admin/DataManager.cfc:498]: There should be a permissions check here. Can also add a `isFlaggingEnabled()` check at the same point.

File: system/handlers/admin/DataManager.cfc [hunk 3]
```diff
@@ -1557,6 +1591,23 @@ component extends="preside.system.base.AdminHandler" {
 			} else {
 				link = event.buildAdminLink( objectName=objectName, operation="editRecord", recordId=recordId );
 			}
+
+			if ( !isEmpty( presideObjectService.getFlagField( objectName ) ) ) {
```
Comment (DominicWatson) [inline system/handlers/admin/DataManager.cfc]: use `presideObjectService.isFlaggingEnabled( objectName )` instead (see review comment further down).

File: system/handlers/admin/DataManager.cfc [hunk 4]
```diff
@@ -1872,6 +1923,7 @@ component extends="preside.system.base.AdminHandler" {
 				var useVersioning   = datamanagerService.isOperationAllowed( objectName, "viewversions" ) && presideObjectService.objectIsVersioned( objectName );
 			}
 
+			var canFlagRecord          = canEdit && !isEmpty( presideObjectService.getFlagField( objectName ) );
```
Comment (DominicWatson) [inline system/handlers/admin/DataManager.cfc]: Use `presideObjectService.isFlaggingEnabled( objectName )` instead (see review comment further down).

**Review by DominicWatson — CHANGES_REQUESTED** [review]
This looks really great. Some minor review comments to address. 

Could you also add a description to the JIRA ticket so that public can know what the feature is all about when reading release notes?

File: system/handlers/admin/DataManager.cfc (hunk 2 above)
Comment (pixl8-brayden) [reply system/handlers/admin/DataManager.cfc:498]: added edit permission check and object flagging enabled check

File: system/services/presideObjects/PresideObjectService.cfc (hunk 1 above)
Comment (pixl8-brayden) [reply system/services/presideObjects/PresideObjectService.cfc:1652]: added

**Comment by DominicWatson** [comment]
Awesome sauce. Merged into upcoming 10.14.0 release.

**Pull request merged**